Rebus.LegacyExtensions/LegacyCompatibility/CustomJsonSerializer.cs

```diff
@@ -0,0 +1,165 @@
+		/// </summary>
+		public const string JsonUtf8ContentType = "application/json;charset=utf-8";
+
+		/// <summary>
+		/// Contents type when the content is JSON
+		/// </summary>
+		public const string JsonContentType = "application/json";
+
+		static readonly JsonSerializerSettings DefaultSettings = new JsonSerializerSettings
+		{
+		   TypeNameHandling = TypeNameHandling.Auto,
+		   Formatting = Formatting.None,
+		   Culture = CultureInfo.InvariantCulture,
+		   Converters = new List<JsonConverter> { new StringEnumConverter() }
+		};
+
+		static readonly Encoding DefaultEncoding = Encoding.UTF8;
+
+		readonly JsonSerializerSettings _settings;
+		readonly Encoding _encoding;
+
+		public CustomJsonSerializer()
+			: this(DefaultSettings, DefaultEncoding)
+		{
+		}
+
+		internal CustomJsonSerializer(Encoding encoding)
+			: this(DefaultSettings, encoding)
+		{
+		}
+
+		internal CustomJsonSerializer(JsonSerializerSettings jsonSerializerSettings)
+			: this(jsonSerializerSettings, DefaultEncoding)
+		{
+		}
+
+		internal CustomJsonSerializer(JsonSerializerSettings jsonSerializerSettings, Encoding encoding)
+		{
+			_settings = jsonSerializerSettings;
+			_encoding = encoding;
+		}
+
+		private string SerializeWithArrayInitialType(object body)
+		{
+			var sb = new StringBuilder(256);
+			var serializer = JsonSerializer.Create(_settings);
+
+			using (StringWriter sw = new StringWriter(sb, CultureInfo.InvariantCulture))
+			using (JsonTextWriter jsonWriter = new JsonTextWriter(sw))
+			{
+				sw.Write("{\"$type\": \"System.Object[]\", \"$values\":");
+				serializer.Serialize(jsonWriter, body);
+				sw.Write("}");
+				return sw.ToString();
+			}
+		}
+
+		/// <summary>
+		/// Serializes the given <see cref="Message"/> into a <see cref="TransportMessage"/>
+		/// </summary>
+		public async Task<TransportMessage> Serialize(Message message)
+		{
+			var jsonText = SerializeWithArrayInitialType(message.Body);
+
+			var bytes = _encoding.GetBytes(jsonText);
+			var headers = message.Headers.Clone();
+			headers[Headers.ContentType] = $"{JsonContentType};charset={_encoding.HeaderName}";
+			return new TransportMessage(headers, bytes);
+		}
+
+		/// <summary>
+		/// Deserializes the given <see cref="TransportMessage"/> back into a <see cref="Message"/>
+		/// </summary>
+		public async Task<Message> Deserialize(TransportMessage transportMessage)
+		{
+			var contentType = transportMessage.Headers.GetValue(Headers.ContentType);
+
+			if (contentType == JsonUtf8ContentType)
+			{
+				return GetMessage(transportMessage, _encoding);
+			}
+
+			if (contentType.StartsWith(JsonContentType))
+			{
+				var encoding = GetEncoding(contentType);
+				return GetMessage(transportMessage, encoding);
+			}
+
+			throw new FormatException($"Unknown content type: '{contentType}' - must be '{JsonUtf8ContentType}' for the JSON serialier to work");
+		}
+
+		Encoding GetEncoding(string contentType)
+		{
+			var charset = contentType
```

Comment:
This is assuming contentType to have some specific format and delimiters, and will fail if/when CT has unexpected values.